Referral program v2

ochaloup opened the pull request:
This is a proposal for changing the deposit instruction. The fucntionality worked for localhost testing.
I would like to kindly ask for feedback about how to enhance and organize the code.

NOTE: my execution of the `./tests.sh` has got issue of not working when running as `+nightly`

luciotato commented on programs/marinade-referral/src/instructions/admin.rs:
```diff
@@ -177,13 +183,44 @@ impl<'info> UpdateReferral<'info> {
         if let Some(new_partner_account) = optional_new_partner_account {
             self.referral_state.partner_account = new_partner_account
         }
+        if let Some(operation_deposit_sol_fee) = self.cap_fee(operation_deposit_sol_fee)? {
```
another option would be:
```
self.referral_state.operation_liquid_unstake_fee = self.checked_fee(operation_liquid_unstake_fee)?;
```

and you check and unwrap in the same fn, returning previous value if None

ochaloup replied:
@luciotato I did in way mentioned in the later comment.

luciotato commented on programs/marinade-referral/src/states.rs:
```diff
@@ -77,6 +77,12 @@ pub struct ReferralState {
     pub pause: bool,
+
+    // fees that will be assigned to referrals, calculated in basis points
+    pub max_operation_fee: u8, // TODO: should this be configurable or a const in program? Does it matter the size of the referral state account?
```
the size of the referral state account does not matter. It is good that it is there and it should be configurable by admin only. Also make it 50 bp by default

More over, when creating the referral state account, add 96 bytes to the account size for future expansion

ochaloup replied:
+1, this will be done on update of the CLI in PR https://github.com/marinade-finance/liquid-staking-referral-cli/pull/2
It's planned for later update when we finalize the fields arrangement in the struct `ReferralState`

ochaloup commented:
I updated the PR with:

* `program`
** I added the field `operation_delayed_unstake_fee`
** the fields started to use `Fee` struct as it implements the functionality of basis point calculations. I was thinking to reuse it.
** the `GlobalState` works with `mint_account` to only later check when refferal account is filled with partner token account. The treasury was removed.
** the `UpdateReferral` instruction verifies the mint and partner owner in the same way as the init does

* `tests`
** moved the tests from separate repository and wanted to refactor it a bit. Just after get it compiling I found I've got an error about double entrypoint after compilation succeeds and tests is about to be started[1]. Is it a wrong idea to refactor this way?


[1] 
```
error: linking with `cc` failed: exit status: 1
  |
...
  = note: /usr/bin/ld: /home/chalda/marinade/liquid-staking-referral-program/target/debug/deps/libmarinade_finance.rlib(marinade_finance.marinade_finance.c69f396e-cgu.0.rcgu.o): in function `entrypoint':
          /home/chalda/.cargo/registry/src/github.com-1ecc6299db9ec823/solana-program-1.7.11/src/entrypoint.rs:46: multiple definition of `entrypoint'; /home/chalda/marinade/liquid-staking-referral-program/target/debug/deps/libmarinade_referral.rlib(marinade_referral.marinade_referral.a06512f1-cgu.15.rcgu.o):/home/chalda/.cargo/registry/src/github.com-1ecc6299db9ec823/solana-program-1.7.11/src/entrypoint.rs:46: first defined here
          collect2: error: ld returned 1 exit status
```

ochaloup commented:
I considered my Rust/Cargo-fu at the low level as I cannot fix the entrypoint error for tests. Thus I moved the tests processing back to state as it was in a separate module (but refractorred to be part of the program repo).
The nonworking entry point tests are at branch: https://github.com/ochaloup/liquid-staking-referral-program/tree/referral-program-v2-not-working-tests

luciotato commented on programs/marinade-referral/src/constant.rs:
```diff
@@ -1,19 +1,13 @@
+///Max % cut for the partner
+pub const DEFAULT_MAX_FEE_POINTS: u32 = 10_000; // 100%
+
+pub const DEFAULT_OPERATION_FEE_POINTS: u32 = 50; // 0.5%
```
Please make DEFAULT_OPERATION_FEE_POINTS = 0

ochaloup commented:
As this PR was approved and I did the requested change (making `DEFAULT_OPERATION_FEE_POINTS = 0`)
https://github.com/marinade-finance/liquid-staking-referral-program/compare/68a24aa71964e1866d0ef830857fc5bd6cfb539f..43c732206e56288541880ff5a4b664a94801ca93
I'm going to merge this and new updates will be part of the new PR.
